### ISSUE TITLE
Performance regression in 3.9.x when updating cached data with identical data

### ISSUE THREAD

**Issue by eltonio450** [apollographql/apollo-client#11645]
### Issue Description

Upgrading from apollo/client 3.8.10 to 3.9.x causes frontend freezing due to a change in behavior when (non)-updating the cache. Specifically, with many cached watch queries, a new query update triggers a complete "reconstruction" of all related observable queries, based on the ids of the received objects (to my understanding), even if the data is identical

In version 3.8.x, if received objects and cached objects properties are identical, no action is taken. 
However, in version 3.9.x, this reconstruction occurs regardless of data similarity. 

With numerous cached queries and large datasets, this exhaustive reconstruction process significantly increases CPU usage, leading to interface freezes. This happens because, even for identical results (e.g., fetching a person named "John Smith" with id = 3), version 3.9.x unnecessarily updates the initial large list, causing about 200ms of CPU activity on a typical machine, as opposed to no activity in version 3.8.10.

### Link to Reproduction

https://github.com/eltonio450/apollo-bug-perf

### Reproduction Steps

Clone the repository and `npm start`(more difficult to reproduce on sandbox bc we need the performance tab in Chrome)
Open the app. Wait a bit
Open performance tab. Run performance monitoring. 
Click on "Add person"
Stop performance monitoring

Do the same with the other version

In 3.8.x, profiling looks like this:
<img width="901" alt="image" src="https://github.com/apollographql/apollo-client/assets/6675365/9f9caabf-203e-4f00-9e21-6301ebea6445">

In 3.9.x, profiling looks like this:
<img width="909" alt="image" src="https://github.com/apollographql/apollo-client/assets/6675365/da152723-e33e-4d13-88da-e9c4fae92f05">

### `@apollo/client` version

3.9.x

**Comment by jerelmiller** [comment]
Hey @eltonio450 👋 

From those screenshots, its a bit difficult to compare those since the 2nd one is a lot more zoomed in than the first. Roughly these look like about the same timing to me (~800ms), but maybe I'm missing something obvious.

That being said, we recently did a lot of work on memory optimization in 3.9. A lot of Apollo's internal caches serve as memoization caches for expensive operations for the internals. Many of these were replaced in 3.9 with LRU caches to avoid memory leaks and with this, the memory limits switched around quite a bit from 3.8. Its possible you're running into a memory limit that you weren't previously in 3.8 given the number of queries you're using at once. This means the client might be doing more recomputation than it was in 3.8.

I'd encourage you to check out the [memory management](https://www.apollographql.com/docs/react/caching/memory-management) doc and play around with the tweaking the memory limits in your app to see if this improves your situation. Perhaps his might help?

If you're curious about a more in-depth background on this work, check out @phryneas's blog post on this: https://www.apollographql.com/blog/apollo-3-9-beta-feature-spotlight-the-memory-story

**Comment by phryneas** [comment]
Yeah, this could be caused by hitting the default limit on `"inMemoryCache.executeSelectionSet"` or `"inMemoryCache.maybeBroadcastWatch"`.

The quickest way you can check for that is calling `__APOLLO_CLIENT__.getMemoryInternals()` in your browser console.

**Comment by eltonio450** [comment]
ohh, thanks! I totally missed those options (although I played with resuktMaxCacheSize)

indeed, inMemoryCache.executeSubSelectedArray was too low. It seems to work now in my example repo, not sure yet if it's the case in prod. I'll let you know

Two questions:
- would it make sense to expose those options when creating the inmemorycache ?
- how do you choose the value? in my app, inMemoryCache.executeSubSelectedArray is around 30k, and not setting it higher has a huuuuge impact (basically it becomes unusable) => would it make sense to have higher default values ? I don't think we run an outlier in terms of memory needs

**Comment by phryneas** [comment]
> would it make sense to expose those options when creating the inmemorycache ?

Not all of these cache values are specific to `InMemoryCache`, and some of them need to be set before you load the Apollo Client bundle (because they concern singleton caches). 
We chose to have one central point of configuration instead of introducing about a dozen different ways of setting the cache - which would not only be more confusing, but also add a lot of bundle size.

> how do you choose the value?

We wrote a blog article, promoted it in various places and, hoped for feedback - but didn't get any. 😓 
Of course, we also tried it out in a bunch of production applications where we took measurements. For what it's worth, the default before was around 65k, so we halved that, but it wasn't *significantly* more.
These cache limits always apply to all the data visible on a single route/screen, and we couldn't find anything that got even close to the new defaults for a single route/screen 
I would assume that in fact your app is an outlier, but of course I could be wrong. Could you explain what kind of data you load here so you manage to hit this limit?

**Comment by eltonio450** [comment]
> > would it make sense to expose those options when creating the inmemorycache ?
> 
> Not all of these cache values are specific to `InMemoryCache`, and some of them need to be set before you load the Apollo Client bundle (because they concern singleton caches). We chose to have one central point of configuration instead of introducing about a dozen different ways of setting the cache - which would not only be more confusing, but also add a lot of bundle size.
> 

maybe for the three that are specific to InMemoryCache ?

> > how do you choose the value?
> 
> We wrote a blog article, promoted it in various places and, hoped for feedback - but didn't get any. 😓 Of course, we also tried it out in a bunch of production applications where we took measurements. For what it's worth, the default before was around 65k, so we halved that, but it wasn't _significantly_ more. These cache limits always apply to all the data visible on a single route/screen, and we couldn't find anything that got even close to the new defaults for a single route/screen I would assume that in fact your app is an outlier, but of course I could be wrong. Could you explain what kind of data you load here so you manage to hit this limit?

I try to read a lot on your work :), it's really impressive btw, but I totally missed those options... I tried sooooo many things (many different typePolicies, completely disabling normalization with `dataIdFromObject: () => false`, etc, etc.), it's quite frustrating to discover those options only now 😅

we have a b2b software, the app I am mentioning is an internal tool with an amount of data I would expect on other internal tools (a few large lists, some big objects, + some notifications and other small objects), but I can never be sure :)

Current value after browsing a bit around on the platform (I must inscrease executeSubSelectedArray)
```
"inMemoryCache": {
      "executeSelectionSet": 30805,
      "executeSubSelectedArray": 10000,
      "maybeBroadcastWatch": 29
    }
```

anyway, it seems to work in prod, closing the issue!

thank you for your help 👍

**Comment by jerelmiller** [comment]
> it's quite frustrating to discover those options only now 😅

Do you have any feedback on how we might have made this a bit more obvious? We tried many of the ways we thought might help (blog, docs, etc.) but perhaps there is something we missed that could have made it more obvious that this was available? Any feedback you can provide would be helpful 🙂 

Glad to hear the increase in memory limits helped though!

**Comment by dylanwulf** [comment]
> Do you have any feedback on how we might have made this a bit more obvious? We tried many of the ways we thought might help (blog, docs, etc.) but perhaps there is something we missed that could have made it more obvious that this was available? Any feedback you can provide would be helpful 🙂

Hi, I'm not involved in this issue but I hope it's ok for me to share my 2 cents here. It's not very obvious from reading the changelog that the blog post or new documentation exist. The changelog is definitely where I would go when upgrading a package to see what's new. Also, that changelog for 3.9.0 is pretty huge and hard to read. The bigger it is, the more likely it is that people will skip reading it entirely and just upgrade hoping that nothing breaks (especially for a minor/patch version). Maybe a summary section or TLDR section in the changelog might make that a little better?
Also, is there a way to subscribe to email updates specifically for Apollo Client blog posts?
I just read through that memory blog post and it's all very impressive work -- thank you guys for working so hard!

**Comment by jerelmiller** [comment]
As always, thanks for your thoughtful feedback 🙂. I'll ask about email updates and let you know!

> Also, that changelog for 3.9.0 is pretty huge and hard to read.

This is really great feedback and we will keep this in mind for future releases!

FWIW, we are switching to a more regular release cadence moving forward so I'm hoping this will solve some of this naturally (shorter time frame = smaller changelog). But the idea about tldr section makes a lot of sense.

Thanks again!

**Comment by jerelmiller** [comment]
@dylanwulf to get back to you on the email updates, it doesn't look like there is a way to subscribe to these posts, but its been noted for the team that owns the blog.

**Comment by anajavi** [comment]
I managed also to hit the same limits. Only realised the root cause by somehow finding this issue.

After increasing the cache sizes, I have:
```
inMemoryCache: 
  executeSelectionSet:  12542
  executeSubSelectedArray: 68802
```

I am guessing these come from having geojson polygon arrays in schema.
